### PR TITLE
Add hold-to-quit button for gameplay/replays

### PR DESCRIPTION
**Pull request by UselessToucan**
I want to be able to exit replay using mouse.
https://www.youtube.com/watch?v=LnXfhS0o_PQ

- [x] Depends on ppy/osu-framework#1560

### CODE REVIEW

**Comment by peppy** [comment]
We really need something like this for play modes too. @arflyte anything in mind?

**Comment by arflyte** [comment]
An "exit" button for play mode and replay?

**Comment by peppy** [comment]
Some mechanism to exit gameplay, for touch users for instance.

**Comment by arflyte** [comment]
Haven't worked on touch interface elements yet. Use this for now, will revisit the buttons again later.

spectator-screen

To quit, player have to hold the button for 1s. "Hold to Quit" fades away after 5s. It's either "Quit" or "Resign", depends on the mode.

On hold, the action icon becomes bigger, "Hold to Quit" is visible again, and progress ring appears.

ee3397bea7

**Comment by peppy** [comment]
Note that this should be displayed in both play and replay, but probably doesn't need to completely fade during replay.

During play it should fade out until the cursor is hovering it and then appear again.

**Comment by UselessToucan** [comment]
Looks ok to me
https://www.youtube.com/watch?v=lg4_EXMZiSs

However I don't know what to do with this new HoldToQuit UI element during actual gameplay.
1. As osu! standard player I just don't need this thing.
2. During gameplay `Exit` call causes player to pause(pause != quit). https://www.youtube.com/watch?v=RiqKzdcun48

**Comment by peppy** [comment]
Where's the test case?

**Comment by UselessToucan** [comment]
Not written yet. :(

**Comment by peppy** [comment]
How did you even make this component without a test case? Please read the wiki regarding how to develop for this project before going any further.

**Comment by peppy** [comment]
This whole thing can be done using transforms and without a clock or whatever.

```csharp

onMouseDown
{
    ring.FillTo(1, duration).OnComplete(_ =>
    {
        // exit
    });
}

onMouseUp
{
    ring.FillTo(0, duration/4); // interrupts the other transform and causes the exit to never occur.
}

```

**Comment by peppy** [comment]
Did you see my comment above? It would clean up the implementation considerably.

**Comment by UselessToucan** [comment]
I saw your comment, but I don't know how to implement this transform. :(

File: osu.Game/Screens/Play/Player.cs [hunk 1]
```diff
@@ -216,6 +216,8 @@ private void load(AudioManager audio, APIAccess api, OsuConfigManager config)
                 }
             };
 
+            hudOverlay.HoldToQuit.Button.ExitAction = Exit;
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/Player.cs]: Would rather that the button wasn't exposed. Make an `Action` property in `HoldToQuit` which proxies the passed value into the button. Make the button itself private.

File: osu.Game/Screens/Play/HUDOverlay.cs [hunk 2]
```diff
@@ -205,6 +207,13 @@ protected virtual ScoreCounter CreateScoreCounter() => new ScoreCounter(6)
             RelativeSizeAxes = Axes.X,
         };
 
+        protected virtual HoldToQuit CreateHoldToQuit() => new HoldToQuit
+        {
+            Anchor = Anchor.BottomRight,
+            Origin = Anchor.BottomRight,
+            Margin = new MarginPadding { Bottom = Progress.Size.Y * 1.25f, Right = 5 }
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/HUDOverlay.cs]: Shouldn't be relative to progress. Just make it 70 or something.

File: osu.Game/Screens/Play/HUD/HoldToQuit.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Graphics.UserInterface;
+using osu.Framework.Input;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using OpenTK;
+
+namespace osu.Game.Screens.Play.HUD
+{
+    public class HoldToQuit : FillFlowContainer
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/HUD/HoldToQuit.cs]: Rename to `QuitButton`.

File: osu.Game/Screens/Play/HUD/HoldToQuit.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                    Font = @"Exo2.0-Bold",
+                    Anchor = Anchor.CentreLeft,
+                    Origin = Anchor.CentreLeft
+                },
+                Button = new HoldToQuitButton(text)
+            };
+            AutoSizeAxes = Axes.Both;
+        }
+
+        public class HoldToQuitButton : CircularContainer
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/HUD/HoldToQuit.cs]: Rename to `Button`, make private.

File: osu.Game/Screens/Play/HUDOverlay.cs [hunk 5]
```diff
@@ -205,6 +207,13 @@ protected virtual ScoreCounter CreateScoreCounter() => new ScoreCounter(6)
             RelativeSizeAxes = Axes.X,
         };
 
+        protected virtual HoldToQuit CreateHoldToQuit() => new HoldToQuit
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/HUDOverlay.cs]: Rename to `CreateQuitButton`.

**Review by smoogipoo — COMMENTED** [review]
Needs some work on the design, but let's go with these reviews for now.

**Review by smoogipoo — CHANGES_REQUESTED** [review]
as commented

**Comment by peppy** [comment]
@UselessToucan highly recommend you check my code changes for this.  look how elegant the shared logic is now

File: osu.Game/Screens/Play/HUD/QuitButton.cs [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        {
+            text.FadeOut(500, Easing.OutQuint);
+            base.OnHoverLost(state);
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            float adjust = Vector2.Distance(GetContainingInputManager().CurrentState.Mouse.NativeState.Position, button.ScreenSpaceDrawQuad.Centre) / 200;
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game/Screens/Play/HUD/QuitButton.cs]: As discussed, let's use ReceiveMouseInputAt() => true here.

File: osu.Game/Screens/Play/HUD/QuitButton.cs [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            double elapsed = MathHelper.Clamp(Clock.ElapsedFrameTime, 0, 1000);
+
+            bool stayVisible = text.Alpha > 0 || button.Progress.Value > 0 || IsHovered;
+
+            Alpha = stayVisible ? 1 : Interpolation.ValueAt(elapsed, Alpha, MathHelper.Clamp(1 - adjust, 0.04f, 1), 0, 200, Easing.OutQuint);
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/HUD/QuitButton.cs]: We'll want a debounce, otherwise it's going to trigger readily for 4:3 players.

Comment (smoogipoo) [reply osu.Game/Screens/Play/HUD/QuitButton.cs]: Also this affects the alpha while the pause screen is active, looks pretty weird.

File: osu.Game/Graphics/Containers/HoldToCofirmContainer.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using System;
+using osu.Framework.Configuration;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+
+namespace osu.Game.Graphics.Containers
+{
+    public abstract class HoldToCofirmContainer : Container
```
Comment (smoogipoo) [inline osu.Game/Graphics/Containers/HoldToCofirmContainer.cs]: HoldToCo **n** firmContainer

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/Player.cs [hunk 9]
```diff
@@ -183,14 +183,14 @@ private void load(AudioManager audio, APIAccess api, OsuConfigManager config)
...
                         },
+                        RulesetContainer.Cursor?.CreateProxy() ?? new Container(),
                         hudOverlay = new HUDOverlay(scoreProcessor, RulesetContainer, working, offsetClock, adjustableClock)
                         {
                             Clock = Clock, // hud overlay doesn't want to use the audio clock directly
                             ProcessCustomClock = false,
                             Anchor = Anchor.Centre,
                             Origin = Anchor.Centre
                         },
-                        RulesetContainer.Cursor?.CreateProxy() ?? new Container(),
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Play/Player.cs]: Yeah, I think it feels better and makes more physical sense

image

File: osu.Game/Screens/Play/HUD/QuitButton.cs [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+                Progress.ValueChanged += v => icon.Scale = new Vector2(1 + (float)v * 0.4f);
+            }
+
+            private bool pendingAnimation;
+
+            protected override void Confirm()
+            {
+                base.Confirm();
+
+                // temporarily unbind as to not look weird during flash animation.
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Play/HUD/QuitButton.cs]: not sure i follow

File: osu.Game/Screens/Play/HUD/QuitButton.cs [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+                Progress.UnbindAll();
+                pendingAnimation = true;
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Play/HUD/QuitButton.cs]: no, it's used to stop a second confirmation while animatng (see it's only usage)

File: osu.Game/Screens/Play/HUD/QuitButton.cs [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+                innerCircle.ScaleTo(0, 100)
+                           .Then().FadeOut().ScaleTo(1).FadeIn(500)
+                           .OnComplete(a => circularProgress.FadeOut(100).OnComplete(_ =>
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Play/HUD/QuitButton.cs]: that's intended (it's not flashing, just a fade out). i guess we should add special behaviour for a user continuing to hold the mouse button down

File: osu.Game/Screens/Play/Player.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game/Screens/Play/Player.cs]: So you're fine with it cursor being below:

1. Leaderboard
2. Key counters

?

I think we need to discuss what the scope of a hud is going to be. Should the hud handle combo pop-in images?

**Comment by peppy** [comment]
All fixed 👍

**Review by smoogipoo — APPROVED** [review]
lgtm

**Pull request merged**